### ISSUE TITLE
Line without measurebars are somehow corrupted

### ISSUE THREAD

**Issue by bwl21** [moinejf/abc2svg#67]
```X:2012
T:testcase 2012
M:9/8
L:1/8
Q:1/4=120
K:G
%%score 1
V:1 clef=treble name="Sopran" snm="S"
cccc
%
[P:Refr.]
c d ccc |
%
```

yields with 1.15.5 

![screenshot_1071](https://user-images.githubusercontent.com/2022976/34326495-3586b96e-e8ae-11e7-938f-417764b6d112.jpg)

while in 1.15.4 it was

![screenshot_1072](https://user-images.githubusercontent.com/2022976/34326497-3dbe94bc-e8ae-11e7-97fe-d3e4614aca71.jpg)

I think the improvements in the line filling results in 

* lines without bars are filled too much to the right. This case happens pretty often so in this case there should be a bit more space on end of line 1
* line 2 being entirely corrupted

**Comment by moinejf** [comment]
I have a fix, but I don't know it raises other problems. May you check it?

```
diff --git a/core/music.js b/core/music.js
index 3beea09..ef1b1be 100644
--- a/core/music.js
+++ b/core/music.js
@@ -4405,9 +4405,10 @@ function set_sym_glue(width) {
                if (s.seqst) {
                        space = s.shrink
                        if (s.space != 0) {
+                               xmax += (s.space > space ? s.space : space) *
+                                               spafac * 1.8
                                if (space < s.space * spafac)
                                        space = s.space * spafac;
-                               xmax += s.space * spafac * 1.8
                        }
                        x += space;
                        xmax += space;
```

**Comment by bwl21** [comment]
I think you can integrate this fix! Thanks.

TL;DR

I have applied the patch and regenerated all my 650 files. 250 of those were different, related to 1.15.4.  Without the patch is was 160 files.

The patch fixed the problem of this issue. I did not observe any dramatic side effect. 
Most (not to say) of the differences are eventually an improvement.

The major source of difference is that we now have a different default for linebreak (which os not the topic of this issue).

**Comment by moinejf** [comment]
Spacing at end of line seems too small. I have an other patch which should be better:

```
diff --git a/core/music.js b/core/music.js
index 3beea09..9ce5deb 100644
--- a/core/music.js
+++ b/core/music.js
@@ -4320,6 +4320,8 @@ function set_piece() {
                last.ts_next = last.next = s;
                s.shrink = last.wr + 8;
                s.space = set_space(s) * .9 - 7
+               if (s.space < s.shrink)
+                       s.space = s.shrink
        }
 }
 
@@ -4404,11 +4406,8 @@ function set_sym_glue(width) {
        while (1) {
                if (s.seqst) {
                        space = s.shrink
-                       if (s.space != 0) {
-                               if (space < s.space * spafac)
-                                       space = s.space * spafac;
-                               xmax += s.space * spafac * 1.8
-                       }
+                       if (s.space != 0)
+                               xmax += s.space * spafac * 1.8;
                        x += space;
                        xmax += space;
                        s.x = x;
```

**Comment by bwl21** [comment]
## looks good

I tried with my testcase (381, the third example above). It is much better. 

![screenshot_1077](https://user-images.githubusercontent.com/2022976/34355706-cbd37076-ea37-11e7-85a2-bd1ece2301e4.jpg)

I played around with it and found that now the distance to the line end is the same as the distance to a bar (if the lined ends with a bar). 

I don't feel really qualified to make comments with respect to an optimum layout for regular notes. So thanks for considering this issue.

## Problems applying the patch

By whatever reason I cannot apply the patches via git ... 

```
git apply --check x.patch
error: patch failed: core/music.js:4320
error: core/music.js: patch does not apply
```

Nevertheless the patch is simple enough such that I could apply it manually. You find my version in a branch of my fork https://github.com/bwl21/abc2svg/tree/issue/67

**Comment by bwl21** [comment]
I have created another testcase:

```abc
X:3058
F:3058_abc3svg-open-line-endings
T:abc3svg open line endings
I:pagewidth 20cm
I:linebreak <EOL>
C:
S:
M:4/4
L:1/4
Q:1/4=120
K:C
%
%
%%score 1 2 3 4
%
V:1 clef=treble name="Sopran" snm="S"
C4 |
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c4 |
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c4 
%
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c c3 |
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c c3 
%
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c2 c2 |
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c2 c2
%
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c3 c |
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c3 c 
%
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c3 c/ c/ |
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c3 c/ c/
%
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c3 c/ c// c// |
c4 | c4 | c4 |c4 | c4 | c4 | c4 | c4 | c4 | c4 | c c/ c// c// 
%
```
Each line is duplicated, with and without bar at the end.

If you play around with this you see that without bar it does not add extra space depending on the duration. This is visible in long notes but not that much in short notes.

![screenshot_1080](https://user-images.githubusercontent.com/2022976/34356741-dfa1416a-ea40-11e7-9930-7f1de60701c1.jpg)

**Comment by moinejf** [comment]
Thanks for the example. Here is the final (?) patch.
A small case is remaining: with whole notes or longer at EOL, spacing is a bit shorter without a bar line.

```
diff --git a/core/music.js b/core/music.js
index 3beea09..78b5892 100644
--- a/core/music.js
+++ b/core/music.js
@@ -2410,7 +2410,8 @@ function set_pitch(last_s) {
                        break
                }
        }
-       smallest_duration = dur
+       if (!last_s)
+               smallest_duration = dur
 }
 
 /* -- set the stem direction when multi-voices -- */
@@ -4318,8 +4319,10 @@ function set_piece() {
                s = add_end_bar(last);
                s.prev = s.ts_prev = last;
                last.ts_next = last.next = s;
-               s.shrink = last.wr + 8;
-               s.space = set_space(s) * .9 - 7
+               s.shrink = last.wr + 2; // just a small space before end of staff
+               s.space = set_space(s)
+               if (s.space < s.shrink)
+                       s.space = s.shrink
        }
 }
 
@@ -4404,11 +4407,8 @@ function set_sym_glue(width) {
        while (1) {
                if (s.seqst) {
                        space = s.shrink
-                       if (s.space != 0) {
-                               if (space < s.space * spafac)
-                                       space = s.space * spafac;
-                               xmax += s.space * spafac * 1.8
-                       }
+                       if (s.space != 0)
+                               xmax += s.space * spafac * 1.8;
                        x += space;
                        xmax += space;
                        s.x = x;
```

**Comment by bwl21** [comment]
Again I had to manually apply the patch ... 

## result for all my files

Now 140 files are different than in 1.15.4 (instead of 250 resp. 160 :-)

there was one strange result (see below)

## result with testcase 3058

* indeed end distance to line end is smaller for whole notes 
* distance in general is a bit smaller than in 1.15.4

![screenshot_1082](https://user-images.githubusercontent.com/2022976/34361798-de2e37c0-ea6d-11e7-8757-2a3955544f97.jpg)

## one strange case

```
X:3040
F:3040_saitenspiel
T:saitenspiel
C:
S:
M:4/4
L:1/4
Q:1/4=120
K:C
%
%
%%score 1
%
V:1 clef=treble octave=-0 name="Sopran" snm="S"
G,CDE | FGAB | cd :|
[P:]
G,/ |: C/D/E/F/ | G A/F/ |E D | C z/ :|
G,/ | C/D/E/F/ | G A/F/ |E D | C z/
[P:]
G/ > F/ E/ G/ | F/E/ D |
G/ > F/ E/ G/ | F/E/ D |
%
C/D/E/F/| G/c/G/F/ | E D | C z
```
Is a strange case but also yields a strange result:

![screenshot_1083](https://user-images.githubusercontent.com/2022976/34362114-fbb4a7c2-ea70-11e7-93b8-152052c20d49.jpg)

**Comment by moinejf** [comment]
My patches were to be applied to the GitHub last commit (92c55dc = tag 1.15.5). There should be no error.
Then, I tried your 'strange case' and I get no error (measures 3 and 7 are full staff, and there is more space at end of the staves without bars).

**Comment by bwl21** [comment]
I tried to apply the patches to tag 1.15.5 but git did not accept.

So i propose that you commit the patches to master. I then will double check.

**Comment by bwl21** [comment]
I have installed commit 32710ea and performed the tests.

Now ALL of my 650 files look different. But as far as I see on the first glance, all are better. I see changes in the horizontal as well as in the vertical positions.

Also "strange case" 3040_saitenspiel looks good.

I need to do some investigations in detail. Nevertheless, i think we are close have this ticket solved :-)

**Comment by bwl21** [comment]
latest regression did not reveal these problems.
